modules/core/02-client/types/params.go

```diff
@@ -22,26 +26,33 @@ func ParamKeyTable() paramtypes.KeyTable {
 }
 
 // NewParams creates a new parameter configuration for the ibc client module
-func NewParams(allowedClients ...string) Params {
+func NewParams(wasmClientAllowed bool, allowedClients ...string) Params {
 	return Params{
-		AllowedClients: allowedClients,
+		WasmClientsEnabled: wasmClientAllowed,
+		AllowedClients:     allowedClients,
```

Comment:
This functionality flag at this moment is not being used by anything (as far as I can tell). To me, it looks like this flag was intended to be a killswitch to prohibit the creation of wasm-based clients (i.e. critical bug found in the module).
@ParthDesai, can you give us a highlight of what this flag aimed to do in the first place?

Overall the auth model seems to be broken here, quick example:
1. `genesis.allowed_clients: {"07-tendermint"}`
1. `simd tx ibc wasm-manager push_wasm dummy_client ...`
2.  This request works:
```
let client_state = ClientState { r#type: "07-tendermint".to_string(), code_id: ... };    

let msg = MsgCreateClient {
  client_state: Some(Any { 
    type_url: "/ibc.lightclients.wasm.v1.ClientState".to_string(), 
    value: prost_serialize(&client_state)?,
});
```

The above example is broken because I register a new client type `dummy_client` and then create an instance of `dummy_client` while specifying `07-tendermint` client type. In short, the codeID is superior to the client type.

I'll analyze it and come back with the fix.

Reply:
The ADR specifies what should happen here

Reply:
@AdityaSripal it's fixed now (https://github.com/cosmos/ibc-go/pull/208/commits/10ea5ced7b9bddc172a31c7e514e5a36272db249).

> Shouldn't WasmClientsEnabled just be encoded inside the allowed clients?

[The ADR](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R54) explains that Wasm light clients are allowed as long as the `WasmClientsEnabled` is true (controlled by governance) and the client type starts with `wasm/` prefix.

Question:
[ICS-24 prohibits](https://github.com/cosmos/ibc-go/blob/db6f316cb044aec21636c2e83f7926ba6c2edfe4/modules/core/24-host/validate.go#L40) `/` character in the client type, and yet the `wasm/` is specified in ADR. To mitigate the issue, I used `wasm-` prefix. Is that okay? Do I need to update the ADR now?

Reply:
@AdityaSripal sorry for a bit of chaos'y comments regarding the client type handling... Hopefully, the latest update is acceptable for everyone.

The [ADR specifies](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R60) that a new Wasm light client type is "dynamically defined" during the light client binary upload.
Example CLI invocation:
```
$ ./simd tx ibc wasm-manager push-wasm "wasm/example-chain" binary.wasm
```

Our initial implementation used the client type as a namespace/alias, where every time the new binary is being uploaded under the same client type, we would version it. For example, the storage key `<client-type>/{latest, previous, next}` would point to selected versions of the binary.

While this looks nice from a user perspective, it raises a governance/ownership question. What prevents "other people" from uploading malicious binary under selected "client-type"? Who owns the "namespace/alias/client-type"?

[This change](https://github.com/cosmos/ibc-go/pull/208/commits/364a2ac1012043f82b3eecf6c72af5c6e19504d8) simplifies the client type management:
1. all Wasm light clients are of `10-wasm` client type. This way we don't have to make any changes to [`IsAllowedClient`](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R60) method
2. every time Wasm blob is uploaded, the caller gets [`code-id` reference](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R81) to uploaded code. This removes the governance requirement.